src/tree/updater_gpu_hist.cu

```diff
@@ -224,6 +224,163 @@ struct CalcWeightTrainParam {
         learning_rate(p.learning_rate) {}
 };
 
+__device__ void atomicOrByte
+(uint* __restrict__ buffer, size_t ibyte, common::CompressedByteT b) {
+  atomicOr(&buffer[ibyte / sizeof(uint)], (uint)b << (ibyte % (sizeof(uint)) * 8));
+}
+
+// index of the first element in cuts greater than v, or n none;
+// cuts are ordered, and binary search is used
+__device__ int upper_bound(const float* __restrict__ cuts, int n, float v) {
+  if (n == 0)
+    return 0;
+  if (cuts[n - 1] <= v)
+    return n;
+  if (cuts[0] > v)
+    return 0;
+  int left = 0, right = n - 1;
+  while (right - left > 1) {
+    int middle = left + (right - left) / 2;
+    if (cuts[middle] > v)
+      right = middle;
+    else
+      left = middle;
+  }
+  return right;
+}
+
+__global__ void max_row_size_k(int* max_size, const size_t* row_ptr, size_t n) {
+  __shared__ int l_max_size;
+  if (threadIdx.x == 0)
+    l_max_size = 0;
+  __syncthreads();
+  size_t i = threadIdx.x + blockIdx.x * blockDim.x;
+  if (i < n)
+    atomicMax(&l_max_size, static_cast<int>(row_ptr[i + 1] - row_ptr[i]));
+  __syncthreads();
+  if (threadIdx.x == 0)
+    atomicMax(max_size, l_max_size);
+}
+
+__global__ void compress_bin_ellpack_k
+(uint* __restrict__ buffer, const size_t* __restrict__ row_ptrs,
+ const RowBatch::Entry* __restrict__ entries,
+ const float* __restrict__ cuts, const size_t* __restrict__ cut_rows,
+ size_t base_row, size_t n_rows, size_t row_ptr_begin, size_t row_stride,
+ int nbits, uint null_gidx_value) {
+  size_t irow = threadIdx.x + size_t(blockIdx.x) * blockDim.x;
+  int ifeature = threadIdx.y + blockIdx.y * blockDim.y;
+  if (irow >= n_rows || ifeature >= row_stride)
+    return;
+  int row_size = static_cast<int>(row_ptrs[irow + 1] - row_ptrs[irow]);
+  uint bin = null_gidx_value;
+  if (ifeature < row_size) {
+    RowBatch::Entry entry = entries[row_ptrs[irow] - row_ptr_begin + ifeature];
+    int feature = entry.index;
+    float fvalue = entry.fvalue;
+    const float *feature_cuts = &cuts[cut_rows[feature]];
+    int ncuts = cut_rows[feature + 1] - cut_rows[feature];
+    bin = upper_bound(feature_cuts, ncuts, fvalue);
+    if (bin >= ncuts)
+      bin = ncuts - 1;
+    bin += cut_rows[feature];
+  }
+  size_t ielement = (irow + base_row) * row_stride + ifeature;
+  size_t ibit_start = ielement * nbits, ibit_end = (ielement + 1) * nbits - 1;
+  size_t ibyte_start = ibit_start / 8, ibyte_end = ibit_end / 8;
+
+  bin <<= 7 - ibit_end % 8;
+  for (ptrdiff_t ibyte = ibyte_end; ibyte >= (ptrdiff_t)ibyte_start; --ibyte) {
+    atomicOrByte(buffer, ibyte, bin & 0xff);
+    bin >>= 8;
+  }
```

Comment:
This bit compression code could be a function somewhere in this file: https://github.com/dmlc/xgboost/blob/master/src/common/compressed_iterator.h

Bit manipulation is extremely prone to errors so I would like to see some unit tests for this.

Reply:
Done.